AsyncDisplayKit/ASVideoNode.mm

```diff
@@ -0,0 +1,352 @@
+}
+
+- (ASButtonNode *)playButton
+{
+  ASDN::MutexLocker l(_lock);
+  
+  return _playButton;
+}
+
+- (void)setAsset:(AVAsset *)asset
+{
+  ASDN::MutexLocker l(_lock);
+  
+  if (ASObjectIsEqual(((AVURLAsset *)asset).URL, ((AVURLAsset *)_asset).URL)) {
+    return;
+  }
+  
+  _asset = asset;
+  
+  // FIXME: Adopt -setNeedsFetchData when it is available
+  if (self.interfaceState & ASInterfaceStateFetchData) {
+    [self fetchData];
+  }
+}
+
+- (AVAsset *)asset
+{
+  ASDN::MutexLocker l(_lock);
+  return _asset;
+}
+
+- (AVPlayer *)player
+{
+  ASDN::MutexLocker l(_lock);
+  return _player;
+}
+
+- (void)setGravity:(ASVideoGravity)gravity
+{
+  ASDN::MutexLocker l(_lock);
+  
+  switch (gravity) {
+    case ASVideoGravityResize:
+      ((AVPlayerLayer *)_playerNode.layer).videoGravity = AVLayerVideoGravityResize;
+      break;
+    case ASVideoGravityResizeAspect:
+      ((AVPlayerLayer *)_playerNode.layer).videoGravity = AVLayerVideoGravityResizeAspect;
+      break;
+    case ASVideoGravityResizeAspectFill:
+      ((AVPlayerLayer *)_playerNode.layer).videoGravity = AVLayerVideoGravityResizeAspectFill;
+      break;
+    default:
+      ((AVPlayerLayer *)_playerNode.layer).videoGravity = AVLayerVideoGravityResizeAspect;
+      break;
+  }
+  
+  _gravity = gravity;
+}
+
+- (ASVideoGravity)gravity
+{
+  ASDN::MutexLocker l(_lock);
+  
+  return _gravity;
+}
+
+#pragma mark - Video Playback
+
+- (void)play
+{
+  ASDN::MutexLocker l(_lock);
+  
+  if (!_spinner) {
+    _spinner = [[ASDisplayNode alloc] initWithViewBlock:^UIView *{
+      UIActivityIndicatorView *spinnnerView = [[UIActivityIndicatorView alloc] init];
+      spinnnerView.color = [UIColor whiteColor];
+      
+      return spinnnerView;
+    }];
+  }
+  
+  [_player play];
+  _shouldBePlaying = YES;
+  _playButton.alpha = 0.0;
+  
+  if (![self ready] && _shouldBePlaying && (self.interfaceState & ASInterfaceStateVisible)) {
+    [self addSubnode:_spinner];
+    [(UIActivityIndicatorView *)_spinner.view startAnimating];
+  }
+}
+
+- (BOOL)ready
+{
+  return _currentItem.status == AVPlayerItemStatusReadyToPlay;
+}
+
+- (void)pause
+{
+  ASDN::MutexLocker l(_lock);
+  
+  [_player pause];
+  [((UIActivityIndicatorView *)_spinner.view) stopAnimating];
+  _shouldBePlaying = NO;
+  _playButton.alpha = 1.0;
+}
+
+- (BOOL)isPlaying
+{
+  ASDN::MutexLocker l(_lock);
+  
+  return (_player.rate > 0 && !_player.error);
+}
+
+#pragma mark - Property Accessors for Tests
+
+- (ASDisplayNode *)spinner
+{
+  ASDN::MutexLocker l(_lock);
+  return _spinner;
+}
+
+- (AVPlayerItem *)curentItem
+{
+  ASDN::MutexLocker l(_lock);
+  return _currentItem;
+}
+
+- (ASDisplayNode *)playerNode
+{
+  ASDN::MutexLocker l(_lock);
+  return _playerNode;
+}
+
+- (BOOL)shouldBePlaying
+{
+  ASDN::MutexLocker l(_lock);
+  return _shouldBePlaying;
+}
+
+#pragma mark - Lifecycle
+
+- (void)dealloc
+{
+  [[NSNotificationCenter defaultCenter] removeObserver:self];
```

Comment:
Since dealloc cost is quite high, let's remove the specific notification we subscribed to.  I only recently learned how much more expensive it is to do the remove-all, as it becomes more expensive the more types and number of notifications / objects are subscribed.